src/Control/Distributed/Process/Internal/CQueue.hs

```diff
@@ -101,33 +106,58 @@ dequeue :: forall m a.
         -> BlockSpec         -- ^ Blocking behaviour
         -> [MatchOn m a]     -- ^ List of matches
         -> IO (Maybe a)      -- ^ 'Nothing' only on timeout
-dequeue (CQueue arrived incoming) blockSpec matchons = mask_ $
+dequeue (CQueue arrived incoming size) blockSpec matchons =
   case blockSpec of
-    Timeout n -> timeout n $ fmap fromJust run
+    Timeout n -> registerDelay n >>= run
     _other    ->
        case chunks of
          [Right ports] -> -- channels only, this is easy:
            case blockSpec of
-             NonBlocking -> atomically $ waitChans ports (return Nothing)
-             _           -> atomically $ waitChans ports retry
+             NonBlocking -> atomically $ decrementJust $ waitChans ports (return Nothing)
+             _           -> atomically $ decrementJust $ waitChans ports retry
                               -- no onException needed
-         _other -> run
+         _other -> newTVarIO False >>= run
   where
     chunks = chunkMatches matchons
 
-    run = do
-           arr <- takeMVar arrived
-           let grabNew xs = do
-                 r <- atomically $ tryReadTChan incoming
-                 case r of
-                   Nothing -> return xs
-                   Just x  -> grabNew (Snoc xs x)
-           arr' <- grabNew arr
-           goCheck chunks arr'
+    decrementJust f = do
+      mx <- f
+      when (isJust mx) (modifyTVar' size pred)
+      return mx
 
     waitChans ports on_block =
         foldr orElse on_block (map (fmap Just) ports)
 
+    run :: TVar Bool -> IO (Maybe a)
+    run tm = do
+        -- We need to wait on both ports and mailbox, here we want
+        -- to break a transaction into multiple ones in order to
+        -- reduce number of retries. Transaction are split into
+        -- parts that are safe to use in concurrent enviroment in
+        -- presence of asynchonous exceptions.
+         
+        -- 1st transaction read consume all values from the input 
+        -- stream. Having this transaction guarantees that all the
+        -- rest computation will not be rolled back if a new value
+        -- will enter the chan.
+	let loop :: IO (IO ())
```

Comment:
I want to compose transaction, i.e. return next transaction outside of current one
(do I need to explain why do I need it, and why it's not possible to do inside one transaction?)
It's possible by wrapping value with a `Foldable` (or at least `Functor`?). Another way
to remove `IO` layer is to use `Maybe` instead, for example:
```
loop :: IO ()
loop = do
   let inner = atomically $ tryReadTChan incoming >>=
        traverse (\x -> modifyTVar' arrived (flip Snoc x) >> return inner)
   forM_ id inner
```